test/001_highsnr_stdio_audio/test_rx.py

```diff
@@ -0,0 +1,180 @@
+
+args = parser.parse_args()
+
+if args.LIST:
+    p = pyaudio.PyAudio()
+    for dev in range(0,p.get_device_count()):
+        print("audiodev: ", dev, p.get_device_info_by_index(dev)["name"])
+    quit()
+   
+N_BURSTS = args.N_BURSTS
+N_FRAMES_PER_BURST = args.N_FRAMES_PER_BURST
+AUDIO_INPUT_DEVICE = args.AUDIO_INPUT_DEVICE
+MODE = codec2.FREEDV_MODE[args.FREEDV_MODE].value
+DEBUGGING_MODE = args.DEBUGGING_MODE
+TIMEOUT = args.TIMEOUT
+
+# AUDIO PARAMETERS
+AUDIO_FRAMES_PER_BUFFER = 2400*2  # <- consider increasing if you get nread_exceptions > 0
+MODEM_SAMPLE_RATE = codec2.api.FREEDV_FS_8000
+AUDIO_SAMPLE_RATE_RX = 48000
+
+# make sure our resampler will work
+assert (AUDIO_SAMPLE_RATE_RX / MODEM_SAMPLE_RATE) == codec2.api.FDMDV_OS_48
+
+# check if we want to use an audio device then do an pyaudio init
+if AUDIO_INPUT_DEVICE != -1: 
+    p = pyaudio.PyAudio()
+    # auto search for loopback devices
+    if AUDIO_INPUT_DEVICE == -2:
+        loopback_list = []
+        for dev in range(0,p.get_device_count()):
+            if 'Loopback: PCM' in p.get_device_info_by_index(dev)["name"]:
+                loopback_list.append(dev)
+        if len(loopback_list) >= 2:
+            AUDIO_INPUT_DEVICE = loopback_list[0] #0  = RX   1 = TX
+            print(f"loopback_list rx: {loopback_list}", file=sys.stderr)
+        else:
+            quit()
+            
+    print(f"AUDIO INPUT DEVICE: {AUDIO_INPUT_DEVICE} DEVICE: {p.get_device_info_by_index(AUDIO_INPUT_DEVICE)['name']}  \
+            AUDIO SAMPLE RATE: {AUDIO_SAMPLE_RATE_RX}", file=sys.stderr)
+    stream_rx = p.open(format=pyaudio.paInt16, 
+                            channels=1,
+                            rate=AUDIO_SAMPLE_RATE_RX,
+                            frames_per_buffer=AUDIO_FRAMES_PER_BUFFER,
+                            input=True,
+                            input_device_index=AUDIO_INPUT_DEVICE
+                            ) 
+
+      
+# ----------------------------------------------------------------
+              
+# DATA CHANNEL INITIALISATION
+
+# open codec2 instance        
+freedv = cast(codec2.api.freedv_open(MODE), c_void_p)
+
+# get number of bytes per frame for mode
+bytes_per_frame = int(codec2.api.freedv_get_bits_per_modem_frame(freedv)/8)
+payload_bytes_per_frame = bytes_per_frame -2
+
+n_max_modem_samples = codec2.api.freedv_get_n_max_modem_samples(freedv)     
+bytes_out = create_string_buffer(bytes_per_frame * 2)
+
+codec2.api.freedv_set_frames_per_burst(freedv,N_FRAMES_PER_BURST)
+
+total_n_bytes = 0
+rx_total_frames = 0
+rx_frames = 0
+rx_bursts = 0
+rx_errors = 0
+nread_exceptions = 0
+timeout = time.time() + TIMEOUT
+receive = True
+audio_buffer = codec2.audio_buffer(AUDIO_FRAMES_PER_BUFFER*2)
+resampler = codec2.resampler()
+
+# Copy received 48 kHz to a file.  Listen to this file with:
+#   aplay -r 48000 -f S16_LE rx48.raw
+# Corruption of this file is a good way to detect audio card issues
+frx = open("rx48.raw", mode='wb')    
```

Comment:
:point_up: monitoring what the Rx receives is how I spotted the audio break up.  I modified one of the tests to play a speech wave file over the virtual card instead of modem tones - that made it really obvious.